File: ts/tests/from_nquads.test.ts
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

import { spawnSync } from "node:child_process";
import { mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "node:test";
import assert from "node:assert/strict";
import { fromNQuads, NQuadsParseError } from "../src/from_nquads.js";
import { Read } from "../src/reader.js";
import { toNQuads } from "../src/nquads.js";
import { RDF_DIR_LANG_STRING, TermKind } from "../src/model.js";
import { Writer } from "../src/writer.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "../../../");
const cli = resolve(__dirname, "../bin/gts.js");
const vectorsDir = join(repoRoot, "vectors");
const RDF_REIFIES = "http://www.w3.org/1999/02/22-rdf-syntax-ns#reifies";

function sortedLines(text: string): string[] {
    return text
        .split(/\r?\n/)
        .map((line) => line.trim())
        .filter((line) => line.length > 0)
        .sort();
}

function roundTrip(nq: string): string {
    return toNQuads(Read(fromNQuads(nq), false));
}

test("fromNQuads inverts fold output for a corpus vector", () => {
    const src = readFileSync(join(vectorsDir, "11-datatype-defaulting.gts"));
    const nq = toNQuads(Read(src, false));
    assert.deepEqual(sortedLines(roundTrip(nq)), sortedLines(nq));
});

test("fromNQuads preserves named graphs, reifiers, and annotations", () => {
    const w = new Writer("dist");
    w.addTerms([
        { kind: TermKind.Iri, value: "https://ex/s" },
        { kind: TermKind.Iri, value: "https://ex/p" },
        { kind: TermKind.Iri, value: "https://ex/o" },
        { kind: TermKind.Iri, value: "https://ex/g" },
        { kind: TermKind.Iri, value: "https://ex/conf" },
        { kind: TermKind.Literal, value: "0.9" },
    ]);
    w.addQuads([{ s: 0, p: 1, o: 2, g: 3 }]);
    w.addReifies([{ rid: 0, spo: { s: 0, p: 1, o: 2 } }]);
    w.addAnnot([{ s: 0, p: 4, o: 5 }]);
    const nq = toNQuads(Read(w.toBytes(), false));
    assert.deepEqual(sortedLines(roundTrip(nq)), sortedLines(nq));
});

test("fromNQuads preserves language-tagged and datatyped literals", () => {
    const xsdInt = "http://www.w3.org/2001/XMLSchema#integer";
    const nq =
        '<https://ex/s> <https://ex/label> "Cat"@en .\n' +
        `<https://ex/s> <https://ex/n> "42"^^<${xsdInt}> .\n` +
        "_:b0 <https://ex/p> <https://ex/s> .\n";
    assert.deepEqual(sortedLines(roundTrip(nq)), sortedLines(nq));
});

test("fromNQuads preserves directional language literals", () => {
    const nq = '<https://ex/s> <https://ex/label> "RTL"@ar--rtl .\n';
    const graph = Read(fromNQuads(nq), false);
    const literal = graph.terms.find((term) => term.kind === TermKind.Literal);
    assert.equal(literal?.lang, "ar");
    assert.equal(literal?.direction, "rtl");
    assert.equal(
        literal === undefined ? undefined : graph.datatypeIri(literal),
        RDF_DIR_LANG_STRING,
    );
    assert.deepEqual(sortedLines(toNQuads(graph)), sortedLines(nq));
});

test("writer allows multiple reifiers for the same statement", () => {
    const w = new Writer("dist");
    w.addTerms([
        { kind: TermKind.Iri, value: "https://ex/r1" },
        { kind: TermKind.Iri, value: "https://ex/r2" },
        { kind: TermKind.Iri, value: "https://ex/s" },
        { kind: TermKind.Iri, value: "https://ex/p" },
        { kind: TermKind.Iri, value: "https://ex/o" },
    ]);
    w.addQuads([{ s: 2, p: 3, o: 4 }]);
    w.addReifies([
        { rid: 0, spo: { s: 2, p: 3, o: 4 } },
        { rid: 1, spo: { s: 2, p: 3, o: 4 } },
    ]);
    const graph = Read(w.toBytes(), false);
    assert.equal(graph.reifiers.length, 2);
    assert.deepEqual(graph.reifiers.map((entry) => entry.rid).sort(), [0, 1]);
});

test("fromNQuads preserves multiple reifiers for the same statement", () => {
    const nq =
        `<https://ex/r1> <${RDF_REIFIES}> <<( <https://ex/s> <https://ex/p> <https://ex/o> )>> .\n` +
        `<https://ex/r2> <${RDF_REIFIES}> <<( <https://ex/s> <https://ex/p> <https://ex/o> )>> .\n`;
    const graph = Read(fromNQuads(nq), false);
    assert.equal(graph.reifiers.length, 2);
    assert.deepEqual(sortedLines(toNQuads(graph)), sortedLines(nq));
});

test("fromNQuads handles compact blank-node and language-tag delimiters", () => {
    const nq =
        "<https://ex/s> <https://ex/p> _:b0.\n" +
        '<https://ex/s> <https://ex/label> "Cat"@en.\n';
    const expected =
        "<https://ex/s> <https://ex/p> _:b0 .\n" +
        '<https://ex/s> <https://ex/label> "Cat"@en .\n';
    assert.deepEqual(sortedLines(roundTrip(nq)), sortedLines(expected));
});

test("fromNQuads keeps quoted-triple close delimiters out of tokens", () => {
    const nq =
        `<https://ex/r1> <${RDF_REIFIES}> <<( _:b0 <https://ex/p> _:b1)>> .\n` +
        `<https://ex/r2> <${RDF_REIFIES}> <<( <https://ex/s> <https://ex/p> "Cat"@en)>> .\n`;
    const expected =
        `<https://ex/r1> <${RDF_REIFIES}> <<( _:b0 <https://ex/p> _:b1 )>> .\n` +
        `<https://ex/r2> <${RDF_REIFIES}> <<( <https://ex/s> <https://ex/p> "Cat"@en )>> .\n`;
    assert.deepEqual(sortedLines(roundTrip(nq)), sortedLines(expected));
});

test("fromNQuads rejects malformed statements", () => {
    assert.throws(
        () => fromNQuads("<https://ex/s> <https://ex/p> .\n"),
        NQuadsParseError,
    );
});

test("fromNQuads rejects empty blank-node labels and language tags", () => {
    assert.throws(
        () => fromNQuads("<https://ex/s> <https://ex/p> _: .\n"),
        NQuadsParseError,
    );
    assert.throws(
        () => fromNQuads('<https://ex/s> <https://ex/p> "Cat"@ .\n'),
        NQuadsParseError,
    );
});

test("CLI from-nq writes a GTS that folds to the input N-Quads", () => {
    const tmp = mkdtempSync(join(tmpdir(), "gts-from-nq-"));
    const src = readFileSync(join(vectorsDir, "11-datatype-defaulting.gts"));
    const nq = toNQuads(Read(src, false));
    const nqPath = join(tmp, "in.nq");
    const outPath = join(tmp, "out.gts");
    writeFileSync(nqPath, nq);
    const result = spawnSync("node", [cli, "from-nq", nqPath, "-o", outPath], {
        encoding: "utf8",
    });
    assert.equal(result.status, 0, result.stderr);
    assert.deepEqual(
        sortedLines(toNQuads(Read(readFileSync(outPath), false))),
        sortedLines(nq),
    );
});

File: go/fromnquads/fromnquads_test.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

package fromnquads

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	"go.blackcatinformatics.ca/gts/model"
	"go.blackcatinformatics.ca/gts/nquads"
	"go.blackcatinformatics.ca/gts/reader"
	"go.blackcatinformatics.ca/gts/wire"
	"go.blackcatinformatics.ca/gts/writer"
	"go.blackcatinformatics.ca/gts/xsd"
)

func sortedLines(text string) []string {
	var lines []string
	for _, line := range strings.Split(strings.TrimSpace(text), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			lines = append(lines, line)
		}
	}
	sort.Strings(lines)
	return lines
}

func vectorsDir(t *testing.T) string {
	t.Helper()
	dir, err := filepath.Abs("../../vectors")
	if err != nil {
		t.Fatal(err)
	}
	return dir
}

func roundTrip(t *testing.T, nq string) string {
	t.Helper()
	data, err := FromNQuads(nq)
	if err != nil {
		t.Fatal(err)
	}
	return nquads.ToNQuads(reader.Read(data, false, nil))
}

func TestFromNQuadsInvertsFoldOutputForCorpusVector(t *testing.T) {
	data, err := os.ReadFile(filepath.Join(vectorsDir(t), "11-datatype-defaulting.gts"))
	if err != nil {
		t.Fatal(err)
	}
	nq := nquads.ToNQuads(reader.Read(data, false, nil))
	if got, want := sortedLines(roundTrip(t, nq)), sortedLines(nq); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func TestFromNQuadsPreservesNamedGraphsReifiersAndAnnotations(t *testing.T) {
	w := writer.New("dist")
	w.AddTerms([]model.Term{
		{Kind: model.Iri, Value: "https://ex/s"},
		{Kind: model.Iri, Value: "https://ex/p"},
		{Kind: model.Iri, Value: "https://ex/o"},
		{Kind: model.Iri, Value: "https://ex/g"},
		{Kind: model.Iri, Value: "https://ex/conf"},
		{Kind: model.Literal, Value: "0.9"},
	})
	gid := 3
	w.AddQuads([]model.Quad{{S: 0, P: 1, O: 2, G: &gid}})
	w.AddReifies([]model.ReifierEntry{{RID: 0, SPO: model.Triple3{S: 0, P: 1, O: 2}, G: &gid}})
	w.AddAnnot([]model.AnnotationEntry{{S: 0, P: 4, O: 5, G: &gid}})
	nq := nquads.ToNQuads(reader.Read(w.ToBytes(), false, nil))
	if got, want := sortedLines(roundTrip(t, nq)), sortedLines(nq); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func TestFromNQuadsPreservesLanguageTaggedAndDatatypedLiterals(t *testing.T) {
	xsdInt := "http://www.w3.org/2001/XMLSchema#integer"
	nq := "<https://ex/s> <https://ex/label> \"Cat\"@en .\n" +
		"<https://ex/s> <https://ex/n> \"42\"^^<" + xsdInt + "> .\n" +
		"_:b0 <https://ex/p> <https://ex/s> .\n"
	if got, want := sortedLines(roundTrip(t, nq)), sortedLines(nq); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func TestFromNQuadsPreservesDirectionalLanguageLiterals(t *testing.T) {
	nq := "<https://ex/s> <https://ex/label> \"RTL\"@ar--rtl .\n"
	graph := reader.Read(mustFromNQuads(t, nq), false, nil)
	var literal *model.Term
	for i := range graph.Terms {
		if graph.Terms[i].Kind == model.Literal {
			literal = &graph.Terms[i]
			break
		}
	}
	if literal == nil {
		t.Fatal("directional literal not found")
	}
	if literal.Lang != "ar" || literal.Direction != "rtl" {
		t.Fatalf("directional literal lost metadata: %#v", literal)
	}
	if got := graph.DatatypeIRI(literal); got != model.RDFDirLangString {
		t.Fatalf("datatype = %q, want %q", got, model.RDFDirLangString)
	}
	if got, want := sortedLines(nquads.ToNQuads(graph)), sortedLines(nq); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func TestFromNQuadsAddsIllTypedLiteralMetadata(t *testing.T) {
	boolIRI := xsd.Namespace + "boolean"
	customIRI := "https://example.org/customDatatype"
	nq := "<https://ex/s> <https://ex/bad> \"maybe\"^^<" + boolIRI + "> .\n" +
		"<https://ex/s> <https://ex/good> \"true\"^^<" + boolIRI + "> .\n" +
		"<https://ex/s> <https://ex/custom> \"not our syntax\"^^<" + customIRI + "> .\n"

	graph := reader.Read(mustFromNQuads(t, nq), false, nil)
	items := xsd.IllTypedLiterals(graph)
	if len(items) != 1 {
		t.Fatalf("ill typed literals = %#v, want one recognized invalid literal", items)
	}
	if diag := items[0].Diagnostic(); diag.Code != xsd.IllTypedLiteralCode {
		t.Fatalf("diagnostic code = %q, want %q", diag.Code, xsd.IllTypedLiteralCode)
	}

	sidecar := metaMap(t, graph, xsd.IllTypedLiteralMetaKey)
	version, ok := wire.AsInt(sidecar["version"])
	if !ok || version != 1 {
		t.Fatalf("sidecar version = %#v, want 1", sidecar["version"])
	}
	rows, ok := sidecar["items"].([]interface{})
	if !ok || len(rows) != 1 {
		t.Fatalf("sidecar items = %#v, want one row", sidecar["items"])
	}
	row, ok := rows[0].(map[interface{}]interface{})
	if !ok {
		t.Fatalf("sidecar row type = %T, want map", rows[0])
	}
	termID, ok := wire.AsInt(row["term"])
	if !ok || termID != items[0].TermID {
		t.Fatalf("row term = %#v, want %d", row["term"], items[0].TermID)
	}
	if got, _ := wire.AsText(row["datatype"]); got != boolIRI {
		t.Fatalf("row datatype = %q, want %q", got, boolIRI)
	}
	if got, _ := wire.AsText(row["lexical"]); got != "maybe" {
		t.Fatalf("row lexical = %q, want maybe", got)
	}
	if got, _ := wire.AsText(row["reason"]); got == "" {
		t.Fatal("row reason is empty")
	}
	if got := nquads.ToNQuads(graph); !strings.Contains(got, "\"not our syntax\"^^<"+customIRI+">") {
		t.Fatalf("unsupported datatype was not preserved in N-Quads:\n%s", got)
	}
}

func TestWriterAllowsMultipleReifiersForSameStatement(t *testing.T) {
	w := writer.New("dist")
	w.AddTerms([]model.Term{
		{Kind: model.Iri, Value: "https://ex/r1"},
		{Kind: model.Iri, Value: "https://ex/r2"},
		{Kind: model.Iri, Value: "https://ex/s"},
		{Kind: model.Iri, Value: "https://ex/p"},
		{Kind: model.Iri, Value: "https://ex/o"},
	})
	w.AddQuads([]model.Quad{{S: 2, P: 3, O: 4}})
	w.AddReifies([]model.ReifierEntry{
		{RID: 0, SPO: model.Triple3{S: 2, P: 3, O: 4}},
		{RID: 1, SPO: model.Triple3{S: 2, P: 3, O: 4}},
	})
	graph := reader.Read(w.ToBytes(), false, nil)
	if len(graph.Reifiers) != 2 {
		t.Fatalf("reifier count = %d, want 2", len(graph.Reifiers))
	}
}

func TestFromNQuadsPreservesMultipleReifiersForSameStatement(t *testing.T) {
	nq := "<https://ex/r1> <" + rdfReifies + "> <<( <https://ex/s> <https://ex/p> <https://ex/o> )>> .\n" +
		"<https://ex/r2> <" + rdfReifies + "> <<( <https://ex/s> <https://ex/p> <https://ex/o> )>> .\n"
	graph := reader.Read(mustFromNQuads(t, nq), false, nil)
	if len(graph.Reifiers) != 2 {
		t.Fatalf("reifier count = %d, want 2", len(graph.Reifiers))
	}
	if got, want := sortedLines(nquads.ToNQuads(graph)), sortedLines(nq); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func TestFromNQuadsCompactBlankNodeAndLanguageTagDelimiters(t *testing.T) {
	nq := "<https://ex/s> <https://ex/p> _:b0.\n" +
		"<https://ex/s> <https://ex/label> \"Cat\"@en.\n"
	expected := "<https://ex/s> <https://ex/p> _:b0 .\n" +
		"<https://ex/s> <https://ex/label> \"Cat\"@en .\n"
	if got, want := sortedLines(roundTrip(t, nq)), sortedLines(expected); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func mustFromNQuads(t *testing.T, nq string) []byte {
	t.Helper()
	data, err := FromNQuads(nq)
	if err != nil {
		t.Fatal(err)
	}
	return data
}

func metaMap(t *testing.T, graph *model.Graph, key string) map[interface{}]interface{} {
	t.Helper()
	for _, entry := range graph.Meta {
		if entry.Key != key {
			continue
		}
		sidecar, ok := entry.Value.(map[interface{}]interface{})
		if !ok {
			t.Fatalf("metadata %s type = %T, want map", key, entry.Value)
		}
		return sidecar
	}
	t.Fatalf("metadata %s not found in %#v", key, graph.Meta)
	return nil
}

func TestFromNQuadsQuotedTripleAdjacentDelimiters(t *testing.T) {
	nq := "<https://ex/r1> <" + rdfReifies + "> <<( _:b0 <https://ex/p> _:b1)>> .\n" +
		"<https://ex/r2> <" + rdfReifies + "> <<( <https://ex/s> <https://ex/p> \"Cat\"@en)>> .\n"
	expected := "<https://ex/r1> <" + rdfReifies + "> <<( _:b0 <https://ex/p> _:b1 )>> .\n" +
		"<https://ex/r2> <" + rdfReifies + "> <<( <https://ex/s> <https://ex/p> \"Cat\"@en )>> .\n"
	if got, want := sortedLines(roundTrip(t, nq)), sortedLines(expected); strings.Join(got, "\n") != strings.Join(want, "\n") {
		t.Fatalf("round-trip mismatch\ngot:\n%s\nwant:\n%s", strings.Join(got, "\n"), strings.Join(want, "\n"))
	}
}

func TestFromNQuadsRejectsMalformedStatements(t *testing.T) {
	if _, err := FromNQuads("<https://ex/s> <https://ex/p> .\n"); err == nil {
		t.Fatal("expected malformed N-Quads to fail")
	}
}

func TestFromNQuadsRejectsEmptyBlankNodeLabelAndLanguageTag(t *testing.T) {
	for _, input := range []string{
		"<https://ex/s> <https://ex/p> _: .\n",
		"<https://ex/s> <https://ex/p> \"Cat\"@ .\n",
	} {
		if _, err := FromNQuads(input); err == nil {
			t.Fatalf("expected malformed N-Quads to fail: %q", input)
		}
	}
}
